src/Workspaces/Core/Portable/Rename/Renamer.cs

```diff
@@ -25,6 +26,96 @@ public static class Renamer
         }
 
+        /// <summary>
+        /// Similar to calling <see cref="Document.WithName(string)" /> with additional changes to the solution. 
+        /// Each change is added as a <see cref="RenameDocumentAction"/> in the returned <see cref="RenameDocumentActionSet.ApplicableActions" />.
+        /// 
+        /// Each action may individually encounter errors that prevent it from behaving correctly. Those are reported in <see cref="RenameDocumentAction.GetErrors(System.Globalization.CultureInfo?)"/>.
+        /// 
+        /// Current supported actions that may be returned: 
+        /// * <see cref="RenameSymbolDocumentAction"/> that will rename the type to match the document name
+        /// </summary>
+        public static async Task<RenameDocumentActionSet> RenameDocumentNameAsync(
+            Document document,
+            string newDocumentName,
+            OptionSet optionSet,
+            CancellationToken cancellationToken = default)
+        {
+            if (document == null)
+            {
+                throw new ArgumentNullException(nameof(document));
+            }
+
+            if (newDocumentName == null)
+            {
+                throw new ArgumentNullException(nameof(newDocumentName));
+            }
+
+            var actions = new ArrayBuilder<RenameDocumentAction>();
+
+            if (!newDocumentName.Equals(document.Name))
+            {
+                var renameAction = await RenameSymbolDocumentAction.TryCreateAsync(document, newDocumentName, optionSet, cancellationToken).ConfigureAwait(false);
+
+                if (renameAction is object)
+                {
+                    actions.Add(renameAction);
+                }
+            }
+
+            return new RenameDocumentActionSet(
+                actions.ToImmutableAndFree(),
+                document.Project.Id,
+                document.Id,
+                newDocumentName,
+                document.Folders);
+        }
+
+        /// <summary>
+        /// Similar to calling <see cref="Document.WithFolders(IEnumerable{string})" /> with additional changes to the solution. 
+        /// Each change is added as a <see cref="RenameDocumentAction"/> in the returned <see cref="RenameDocumentActionSet.ApplicableActions" />.
+        /// 
+        /// Each action may individually encounter errors that prevent it from behaving correctly. Those are reported in <see cref="RenameDocumentAction.GetErrors(System.Globalization.CultureInfo?)"/>.
+        /// 
+        /// Current supported actions that may be returned: 
+        /// * <see cref="SyncNamespaceDocumentAction"/> that will sync the namespace(s) of the document to match the document folders
+        /// </summary>
+        public static async Task<RenameDocumentActionSet> RenameDocumentFoldersAsync(
+            Document document,
+            IReadOnlyList<string> newFolders,
+            OptionSet optionSet,
+            CancellationToken cancellationToken = default)
```

Comment:
i don't get the point of hte action set.  in both entrypoints we're only returning a single action.  so why not just return that action.  what is the set for?

Reply:
We have plans to add more actions. This design supports that plan

Reply:
can you LMK what other actions we intend to add?  i'm very wary about speculative need here without a crisp understanding of how we woudl use this.

Reply:
Modify header